### ISSUE TITLE
Custom handler file with multiple classes: not selecting the first one

### ISSUE THREAD

**Issue by DSLituiev** [pytorch/serve#964]
According to [docs](https://github.com/pytorch/serve/blob/master/docs/custom_service.md#custom-handler-with-class-level-entry-point), if two classes are present in a handler file, the first one will be used. However, that's not what happens, and I get an error while serving. Additionally, specifying the handler class with `--handler handler.py:HandlerClass` yields an error during archival.

I have two classes in the `handler.py` : `ClassifierFeatureExtractor` and `ScaleIntensityTransform`

 ## attempt 1: archiving with `--handler handler.py`
- archival works w/o errors
- server stumbles:
```
ValueError: Expected only one class in custom service code or a function entry point
 [<class 'baidu_handler.ClassifierFeatureExtractor'>, <class 'baidu_handler.ScaleIntensityTransform'>]
```

 ## attempt 2: archiving with `--handler handler.py:ClassifierFeatureExtractor`
- archival fails:
```
FileNotFoundError: [Errno 2] No such file or directory: 'handler.py:ClassifierFeatureExtractor'
```

## Context

```
pip freeze |grep torch

torch==1.7.1
torch-model-archiver==0.3.0
torchserve==0.3.0
torchvision==0.2.0

```
* java version
```
java --version
openjdk 11.0.9.1 2020-11-04
OpenJDK Runtime Environment (build 11.0.9.1+1-Ubuntu-0ubuntu1.20.04)
OpenJDK 64-Bit Server VM (build 11.0.9.1+1-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)
```

* Operating System and version: ubuntu.20.04

## Your Environment

* Installed using source? [yes/no]: pip
* Are you planning to deploy it using docker container? [yes/no]: NA
* Is it a CPU or GPU environment?: GPU 
* Using a default/custom handler? [If possible upload/share custom handler/model]: see above
* What kind of model is it e.g. vision, text, audio?: vision

...

## Failure Logs [if any]

<details>
```
2021-02-04 10:39:11,362 [INFO ] main org.pytorch.serve.ModelServer -
Torchserve version: 0.3.0
TS Home: ~/.conda/envs/ncrf/lib/python3.6/site-packages
Current directory: ~/repos/NCRF/scripts
Temp directory: /tmp
Number of GPUs: 1
Number of CPUs: 8
Max heap size: 3986 M
Python executable: ~/.conda/envs/ncrf/bin/python
Config file: logs/config/20210203222847767-shutdown.cfg
Inference address: http://127.0.0.1:8080
Management address: http://127.0.0.1:8081
Metrics address: http://127.0.0.1:8082
Model Store: ~/repos/NCRF/model_store
Initial Models: resnet18-baseline=resnet18-baseline.mar
Log dir: ~/repos/NCRF/scripts/logs
Metrics dir: ~/repos/NCRF/scripts/logs
Netty threads: 0
Netty client threads: 0
Default workers per model: 1
Blacklist Regex: N/A
Maximum Response Size: 6553500
Maximum Request Size: 6553500
Prefer direct buffer: false
Allowed Urls: [file://.*|http(s)?://.*]
Custom python dependency for model allowed: false
Metrics report format: prometheus
Enable metrics API: true

2021-02-04 10:39:11,373 [INFO ] main org.pytorch.serve.snapshot.SnapshotManager - Started restoring models from snapshot {
  "name": "20210203222847767-shutdown.cfg",
  "modelCount": 1,
  "created": 1612420127768,
  "models": {
    "resnet18-baseline": {
      "1.0": {
        "defaultVersion": true,
        "marName": "resnet18-baseline.mar",
        "minWorkers": 1,
        "maxWorkers": 1,
        "batchSize": 1,
        "maxBatchDelay": 100,
        "responseTimeout": 120
      }
    }
  }
}
2021-02-04 10:39:11,385 [INFO ] main org.pytorch.serve.snapshot.SnapshotManager - Validating snapshot 20210203222847767-shutdown.cfg
2021-02-04 10:39:11,386 [INFO ] main org.pytorch.serve.snapshot.SnapshotManager - Snapshot 20210203222847767-shutdown.cfg validated successfully
2021-02-04 10:39:11,875 [INFO ] main org.pytorch.serve.archive.ModelArchive - eTag 83d761631a2943f08c5764350f3ebd76
2021-02-04 10:39:11,883 [DEBUG] main org.pytorch.serve.wlm.ModelVersionedRefs - Adding new version 1.0 for model resnet18-baseline
2021-02-04 10:39:11,884 [DEBUG] main org.pytorch.serve.wlm.ModelVersionedRefs - Setting default version to 1.0 for model resnet18-baseline
2021-02-04 10:39:11,884 [DEBUG] main org.pytorch.serve.wlm.ModelVersionedRefs - Setting default version to 1.0 for model resnet18-baseline
2021-02-04 10:39:11,884 [INFO ] main org.pytorch.serve.wlm.ModelManager - Model resnet18-baseline loaded.
2021-02-04 10:39:11,884 [DEBUG] main org.pytorch.serve.wlm.ModelManager - updateModel: resnet18-baseline, count: 1
2021-02-04 10:39:11,904 [INFO ] main org.pytorch.serve.ModelServer - Initialize Inference server with: EpollServerSocketChannel.
2021-02-04 10:39:11,954 [INFO ] main org.pytorch.serve.ModelServer - Inference API bind to: http://127.0.0.1:8080
2021-02-04 10:39:11,955 [INFO ] main org.pytorch.serve.ModelServer - Initialize Management server with: EpollServerSocketChannel.
021-02-04 10:39:11,904 [INFO ] main org.pytorch.serve.ModelServer - Initialize Inference server with: EpollServerSocketChannel.
2021-02-04 10:39:11,954 [INFO ] main org.pytorch.serve.ModelServer - Inference API bind to: http://127.0.0.1:8080
2021-02-04 10:39:11,955 [INFO ] main org.pytorch.serve.ModelServer - Initialize Management server with: EpollServerSocketChannel.
2021-02-04 10:39:11,955 [INFO ] main org.pytorch.serve.ModelServer - Management API bind to: http://127.0.0.1:8081
2021-02-04 10:39:11,955 [INFO ] main org.pytorch.serve.ModelServer - Initialize Metrics server with: EpollServerSocketChannel.
2021-02-04 10:39:11,956 [INFO ] main org.pytorch.serve.ModelServer - Metrics API bind to: http://127.0.0.1:8082
2021-02-04 10:39:11,986 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - Listening on port: /tmp/.ts.sock.9000
2021-02-04 10:39:11,986 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - [PID]2460
2021-02-04 10:39:11,986 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - Torch worker started.
2021-02-04 10:39:11,987 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - Python runtime: 3.6.12
2021-02-04 10:39:11,987 [DEBUG] W-9000-resnet18-baseline_1.0 org.pytorch.serve.wlm.WorkerThread - W-9000-resnet18-baseline_1.0 State change null -> WORKER_STARTED
2021-02-04 10:39:11,997 [INFO ] W-9000-resnet18-baseline_1.0 org.pytorch.serve.wlm.WorkerThread - Connecting to: /tmp/.ts.sock.9000
2021-02-04 10:39:12,005 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - Connection accepted: /tmp/.ts.sock.9000.
2021-02-04 10:39:13,935 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - Backend worker process died.
2021-02-04 10:39:13,936 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - Traceback (most recent call last):
2021-02-04 10:39:13,936 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -   File "~/.conda/envs/ncrf/lib/python3.6/site-packages/ts/model_service_worker.py", line 182, in <module>
2021-02-04 10:39:13,936 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -     worker.run_server()
2021-02-04 10:39:13,936 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -   File "~/.conda/envs/ncrf/lib/python3.6/site-packages/ts/model_service_worker.py", line 154, in run_server
2021-02-04 10:39:13,936 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -   File "~/.conda/envs/ncrf/lib/python3.6/site-packages/ts/model_service_worker.py", line 154, in run_server
2021-02-04 10:39:13,936 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -     self.handle_connection(cl_socket)
2021-02-04 10:39:13,936 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -   File "~/.conda/envs/ncrf/lib/python3.6/site-packages/ts/model_service_worker.py", line 116, in handle_connection
2021-02-04 10:39:13,936 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -     service, result, code = self.load_model(msg)
2021-02-04 10:39:13,936 [INFO ] epollEventLoopGroup-5-1 org.pytorch.serve.wlm.WorkerThread - 9000 Worker disconnected. WORKER_STARTED
2021-02-04 10:39:13,936 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -   File "~/.conda/envs/ncrf/lib/python3.6/site-packages/ts/model_service_worker.py", line 89, in load_model
2021-02-04 10:39:13,937 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -     service = model_loader.load(model_name, model_dir, handler, gpu, batch_size, envelope)
2021-02-04 10:39:13,937 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -   File "~/.conda/envs/ncrf/lib/python3.6/site-packages/ts/model_loader.py", line 96, in load
2021-02-04 10:39:13,937 [DEBUG] W-9000-resnet18-baseline_1.0 org.pytorch.serve.wlm.WorkerThread - System state is : WORKER_STARTED
2021-02-04 10:39:13,937 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -     entry_point, initialize_fn = self._get_class_entry_point(module)
2021-02-04 10:39:13,937 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -   File "~/.conda/envs/ncrf/lib/python3.6/site-packages/ts/model_loader.py", line 138, in _get_class_entry_point
2021-02-04 10:39:13,937 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle -     model_class_definitions))
2021-02-04 10:39:13,937 [DEBUG] W-9000-resnet18-baseline_1.0 org.pytorch.serve.wlm.WorkerThread - Backend worker monitoring thread interrupted or backend worker process died.
java.lang.InterruptedException
        at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.reportInterruptAfterWait(AbstractQueuedSynchronizer.java:2056)
        at java.base/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2133)
                at java.base/java.util.concurrent.ArrayBlockingQueue.poll(ArrayBlockingQueue.java:432)
        at org.pytorch.serve.wlm.WorkerThread.run(WorkerThread.java:188)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
2021-02-04 10:39:13,937 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - ValueError: Expected only one class in custom service code or a function entry point [<class 'baidu_handler.BaiduClassifierFeatureExtractor'>, <class 'baidu_handler.ScaleIntensityBaidu'>]
2021-02-04 10:39:13,939 [WARN ] W-9000-resnet18-baseline_1.0 org.pytorch.serve.wlm.BatchAggregator - Load model failed: resnet18-baseline, error: Worker died.
2021-02-04 10:39:13,939 [DEBUG] W-9000-resnet18-baseline_1.0 org.pytorch.serve.wlm.WorkerThread - W-9000-resnet18-baseline_1.0 State change WORKER_STARTED -> WORKER_STOPPED
2021-02-04 10:39:13,939 [WARN ] W-9000-resnet18-baseline_1.0 org.pytorch.serve.wlm.WorkerLifeCycle - terminateIOStreams() threadName=W-9000-resnet18-baseline_1.0-stderr
2021-02-04 10:39:13,939 [WARN ] W-9000-resnet18-baseline_1.0 org.pytorch.serve.wlm.WorkerLifeCycle - terminateIOStreams() threadName=W-9000-resnet18-baseline_1.0-stdout
2021-02-04 10:39:13,940 [INFO ] W-9000-resnet18-baseline_1.0 org.pytorch.serve.wlm.WorkerThread - Retry worker: 9000 in 1 seconds.
2021-02-04 10:39:13,962 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - Stopped Scanner - W-9000-resnet18-baseline_1.0-stdout
2021-02-04 10:39:13,962 [INFO ] W-9000-resnet18-baseline_1.0-stderr org.pytorch.serve.wlm.WorkerLifeCycle - Stopped Scanner - W-9000-resnet18-baseline_1.0-stderr
2021-02-04 10:39:15,013 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - Listening on port: /tmp/.ts.sock.9000
2021-02-04 10:39:15,013 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - [PID]2490
2021-02-04 10:39:15,013 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - Torch worker started.
2021-02-04 10:39:15,013 [INFO ] W-9000-resnet18-baseline_1.0-stdout org.pytorch.serve.wlm.WorkerLifeCycle - Python runtime: 3.6.12
2021-02-04 10:39:15,013 [DEBUG] W-9000-resnet18-baseline_1.0 org.pytorch.serve.wlm.WorkerThread - W-9000-resnet18-baseline_1.0 State change WORKER_STOPPED -> WORKER_STARTED
```
</details>

**Comment by maaquib** [comment]
@DSLituiev 
* Issue 1: This seems to be an issue with the docs being outdated. Will update
* Issue 2: Can you try archiving with `--handler handler:ClassifierFeatureExtractor` and let me know if it works https://github.com/pytorch/serve/tree/master/model-archiver#handler

**Comment by DSLituiev** [comment]
method 2  worked, at least I don't see an error. dealing with other errors now

**Comment by DSLituiev** [comment]
There still might be a remaining issue here:

```
File "/home/dlituiev/.conda/envs/ncrf/lib/python3.6/site-packages/ts/model_loader.py", line 131, in <lambda>
initialize_fn = lambda ctx: entry_point(None, ctx)
TypeError: __init__() takes 1 positional argument but 3 were given
```

I inserted a log statement after line 130 in `ts/model_loader.py`:

```
    def _get_function_entry_point(self, module, function_name):
        logging.info(f"initializing module {module}.{function_name}()")
        entry_point = getattr(module, function_name)
        initialize_fn = lambda ctx: entry_point(None, ctx)
        return entry_point, initialize_fn
```
Output:
```initializing module <module 'original_baidu_handler' from '/tmp/models/6c7b776e96b84c56be6a773281e62ce3/original_baidu_handler.py'>.BaiduClassifierFeatureExtractor()```

Which makes me think it treats the class as if it were a function

**Comment by DSLituiev** [comment]
specifying `--handler handler:ClassifierFeatureExtractor.handle` brings me back to the same error:
```
ValueError: Expected only one class in custom service code or a function entry point
 [<class 'baidu_handler.ClassifierFeatureExtractor'>, <class 'baidu_handler.ScaleIntensityTransform'>]
```

**Comment by maaquib** [comment]
@DSLituiev As I mentioned for the first issue, the docs are outdated. Looks like TS [doesn't support](https://github.com/pytorch/serve/blob/abcaf1a062dd2dd60352955bdd5ecfbf08cf3b47/ts/model_loader.py#L136) multiple classes in the handler right now. Would it be possible to reorganize your code to have only a single class in handler.py?

**Comment by DSLituiev** [comment]
I didn't have a function as entry point initially. I though a class would do the job from docs.

Now I wrapped the class into a function
```
def handle(data, context):
    obj = ClassifierFeatureExtractor()
    return obj.handle(data, context)
```

and compiled with `--handler  handler:handle`.

Still getting a tons of logs that are really hard to parse by eye and it keeps restarting. Is there a way to tell it to exit on error? Or any other way to separate Python logs from the rest?

**Comment by DSLituiev** [comment]
I keep seeing messages "File xyz.py in fuction blahblah", but I can't see the actual exception.

**Comment by maaquib** [comment]
Would it be possible for you to share your handler code? If not I'll try and code up something similar and see if I can reproduce the issue.

**Comment by DSLituiev** [comment]
What I see is that the handler is called when I start the server and before I pass any request:
`torchserve --start --model-store model_store --models resnet18-baseline=esnet18-baseline.mar`

I see the server fails in  [`ts/model_loader.py:104`](https://github.com/pytorch/serve/blob/master/ts/model_loader.py#L104), while calling   `initialize_fn(service.context)`. The latter is defined as `initialize_fn = lambda ctx : handler(None, ctx)` [in `_get_function_entry_point`](https://github.com/pytorch/serve/blob/master/ts/model_loader.py#L131). This looks surprising, as I my handler is a subclass of `BaseHandler` that [passes](https://github.com/pytorch/serve/blob/master/ts/torch_handler/base_handler.py#L194) the data to convert `torch.tensor(data)` and fails with the `None` input. Thus I am surprised to see that my custom handler is called differently from how I would expect `BasicHandler`s would be called.

Given examples that you provide, like [here](https://github.com/pytorch/serve/blob/master/docs/custom_service.md#advanced-custom-handlers), handler function seem to expect proper data, and not take special care of `None` case. Am I missing something?

**Comment by DSLituiev** [comment]
Thank you for looking into it.
Her is the [code](https://gist.github.com/DSLituiev/4d4bbf51a8bbdcd46d6b24f0fa76e15a)

**Comment by DSLituiev** [comment]
I was able to resolve it by moving the transform class into a separate file and passing the latter under the `--extra-files`.

**Comment by lxning** [comment]
@DSLituiev I'm going to close this bug. Please feel free to reopen it if you find new issues.

**Comment by DSLituiev** [comment]
@lxning I believe this still needs to be documented properly.

**Comment by MichaelFishmanBDAII** [comment]
The [documentation](https://pytorch.org/serve/custom_service.html#custom-handler-with-class-level-entry-point) still says we can have multiple classes in the handler file, as long as the handler class is the first one.
